Cython/Utility/CommonStructures.c

```diff
@@ -7,90 +7,109 @@ static PyObject* __Pyx_FetchCommonTypeFromSpec(PyType_Spec *spec, PyObject *base
-            (PyObject *)cached_type,
-            type->tp_name,
-            cached_type->tp_basicsize,
-            type->tp_basicsize);
+        if (__Pyx_VerifyCachedType(
+              (PyObject *)cached_type,
+              type->tp_name,
+              cached_type->tp_basicsize,
+              type->tp_basicsize) < 0) {
+            goto bad;
+        }
         goto done;
     }
 
-    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto done;
+    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto bad;
     PyErr_Clear();
-    if (PyType_Ready(type) < 0) goto done;
-    if (PyObject_SetAttrString(fake_module, type->tp_name, (PyObject *)type) < 0)
-        goto done;
+    if (PyType_Ready(type) < 0) goto bad;
+    if (PyObject_SetAttrString(abi_module, type->tp_name, (PyObject *)type) < 0)
+        goto bad;
     Py_INCREF(type);
     cached_type = type;
 
 done:
+    Py_XDECREF(abi_module);
     // NOTE: always returns owned reference, or NULL on error
-    Py_XDECREF(fake_module);
-    Py_XDECREF(cached_type);
     return cached_type;
+
+bad:
+    Py_XDECREF(cached_type);
+    cached_type = NULL;
+    goto done;
 }
 
 #if CYTHON_COMPILING_IN_LIMITED_API
 static PyObject *__Pyx_FetchCommonTypeFromSpec(PyType_Spec *spec, PyObject *bases) {
-    PyObject *fake_module, *py_basicsize, *type, *cached_type = NULL;
+    PyObject *abi_module, *py_basicsize, *type, *cached_type = NULL;
     Py_ssize_t basicsize;
 
-    fake_module = fetch_fake_module();
-    if (!fake_module) goto done;
-    cached_type = PyObject_GetAttrString(fake_module, spec->name);
+    abi_module = __Pyx_FetchSharedCythonABIModule();
+    if (!abi_module) goto bad;
+    cached_type = PyObject_GetAttrString(abi_module, spec->name);
     if (cached_type) {
         py_basicsize = PyObject_GetAttrString(cached_type, "__basicsize__");
-        if (!py_basicsize) goto done;
+        if (!py_basicsize) goto bad;
         basicsize = PyLong_AsSsize_t(py_basicsize);
         Py_DECREF(py_basicsize);
         py_basicsize = 0;
-        if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred()) goto done;
-        verify_cached_type(cached_type, spec->name, basicsize, spec->basicsize);
+        if (basicsize == (Py_ssize_t)-1 && PyErr_Occurred()) goto bad;
+        if (__Pyx_VerifyCachedType(
+              cached_type,
+              spec->name,
+              basicsize,
+              spec->basicsize) < 0) {
+            goto bad;
+        }
         goto done;
     }
 
-    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto done;
+    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto bad;
     PyErr_Clear();
     type = PyType_FromSpecWithBases(spec, bases);
-    if (unlikely(!type)) goto done;
-    if (PyObject_SetAttrString(fake_module, spec->name, type) < 0) goto done;
+    if (unlikely(!type)) goto bad;
+    if (PyObject_SetAttrString(abi_module, spec->name, type) < 0) goto bad;
     cached_type = type;
 
 done:
+    Py_XDECREF(abi_module);
```

Comment:
`Py_DECREF()` is enough here, it's always non-NULL.

Reply:
Done!